teps/0115-tekton-catalog-git-based-versioning.md

```diff
@@ -0,0 +1,398 @@
+For example, the directory structure with two `Tasks` and one `Pipeline` (each with two versions) is organized as such:
+
+```shell
+./task/
+  /argocd
+    /0.1
+      /README.md
+      /argocd.yaml
+      /samples
+      /tests
+    /0.2
+      /README.md
+      /argocd.yaml
+      /samples
+      /tests
+    /OWNERS
+  /golang-build
+    /0.1
+      /README.md
+      /golang-build.yaml
+      /samples
+      /tests
+    /0.2
+      /README.md
+      /golang-build.yaml
+      /samples
+      /tests
+    /OWNERS
+./pipeline/
+  /go-release
+    /0.1
+      /README.md
+      /go-release.yaml
+      /samples
+      /tests
+    /0.2
+      /README.md
+      /go-release.yaml
+      /samples
+      /tests
+    /OWNERS
+```
+
+For further information, see [TEP-0003: Tekton Catalog Organization][tep-0003].
+
+The current directory-based organization and versioning presents challenges:
+- Bumping versions of resources is cumbersome which encourages making "minor" changes in existing versions. Some changes
+that seem minor cause surprising breaking changes for users, such as described in [tektoncd/catalog#784][784].
+- Resources that are frequently updated with new versions end up with a lot of duplications because any change requires
+copying over everything into a new directory. For example, [git-clone][git-clone] `Task` has seven versions with a lot
+of duplication across versions.
+
+Previously, there was tight coupling between the organization and how users referenced resources which made it hard to 
+change the organization of the Catalog without breaking users. In [TEP-0110][tep-0110], we decoupled the Catalog 
+organization from resource reference. Now, we can create a new organization and versioning contract for Catalogs to
+address the challenges with the directory-based organization and versioning.
+
+The aim of this TEP is to create a new organization and versioning for Tekton Catalogs, building on the prior work and 
+experiences from [TEP-0003][tep-0003]. In [TEP-0079: Tekton Catalog Support Tiers][tep-0079], we plan to add a Tekton 
+Official Catalog which will use this new organization and versioning contract - therefore, this proposal is a blocker 
+for [TEP-0079][tep-0079].
+
+## Proposal
+
+We propose supporting a flat organization and git-based versioning of Tekton Catalogs.
+
+### Flat Organization
+
+In this organization contract, we'll use a flat structure instead of the directory structure. The example shown above
+with the two `Tasks` (argocd and golang-build) and one `Pipeline` (go-release) would be organized as such:
+
+```shell
+./task/
+  /argocd
+    /README.md
+    /argocd.yaml
+    /samples
+    /tests
+    /OWNERS
+  /golang-build
+    /README.md
+    /golang-build.yaml
+    /samples
+    /tests
+    /OWNERS
+./pipeline/
+  /go-release
+    /README.md
+    /go-release.yaml
+    /samples
+    /tests
+    /OWNERS
+```
```

Comment:
This looks it would be a good organisation only if the `argocd` and `golang-build` tasks where only ever used by the `go-release` pipeline. If that is the case they could as well be embedded in the pipeline spec.